table/pit-cs-tree.go

```diff
@@ -0,0 +1,382 @@
+
+	var entry *nameTreePitEntry
+	for _, curEntry := range node.pitEntries {
+		if curEntry.CanBePrefix() == interest.CanBePrefix() && curEntry.MustBeFresh() == interest.MustBeFresh() && ((hint == nil && curEntry.ForwardingHint == nil) || hint.Equals(curEntry.ForwardingHint())) {
+			entry = curEntry
+			break
+		}
+	}
+
+	if entry == nil {
+		p.nPitEntries++
+		entry = new(nameTreePitEntry)
+		entry.node = node
+		entry.pitCsTable = p
+		entry.name = interest.Name()
+		entry.canBePrefix = interest.CanBePrefix()
+		entry.mustBeFresh = interest.MustBeFresh()
+		entry.forwardingHint = hint
+		entry.inRecords = make(map[uint64]*PitInRecord)
+		entry.outRecords = make(map[uint64]*PitOutRecord)
+		entry.satisfied = false
+		node.pitEntries = append(node.pitEntries, entry)
+		entry.token = p.generateNewPitToken()
+		p.pitTokenMap[entry.token] = entry
+	}
+
+	for face, inRecord := range entry.inRecords {
+		// Only considered a duplicate (loop) if from different face since is just retransmission and not loop if same face
+		if face != inFace && bytes.Equal(inRecord.LatestNonce, interest.Nonce()) {
+			return entry, true
+		}
+	}
+
+	// Cancel expiration time
+	entry.expirationTime = time.Unix(0, 0)
+
+	return entry, false
+}
+
+// RemovePITEntry removes the specified PIT entry.
+// Originally implemented as RemovePITEntry()
+func (p *PitCsTree) RemoveInterest(pitEntry PitEntry) bool {
+	e := pitEntry.(*nameTreePitEntry)
+	for i, entry := range e.node.pitEntries {
+		if entry == pitEntry {
+			if i < len(e.node.pitEntries)-1 {
+				copy(e.node.pitEntries[i:], e.node.pitEntries[i+1:])
+			}
+			e.node.pitEntries = e.node.pitEntries[:len(e.node.pitEntries)-1]
+			if len(e.node.pitEntries) == 0 {
+				entry.node.pruneIfEmpty()
+			}
+			p.nPitEntries--
+			return true
+		}
+	}
+	return false
+}
+
+// Originally implemented as FindOrInsertPIT()
+func (p *PitCsTree) FindInterestExactMatch(interest *ndn.Interest) PitEntry {
+	node := p.root.findExactMatchEntry(interest.Name())
+	for _, curEntry := range node.pitEntries {
+		if curEntry.CanBePrefix() == interest.CanBePrefix() && curEntry.MustBeFresh() == interest.MustBeFresh() {
+			return curEntry
+		}
+	}
+	return nil
+}
+
+// Logic taken from FindPITFromData
+func (p *PitCsTree) FindInterestPrefixMatch(interest *ndn.Interest, token uint32) []PitEntry {
+	matching := make([]PitEntry, 0)
+	dataNameLen := interest.Name().Size()
+	for curNode := p.root.findLongestPrefixEntry(interest.Name()); curNode != nil; curNode = curNode.parent {
+		for _, entry := range curNode.pitEntries {
+			if entry.canBePrefix || curNode.depth == dataNameLen {
+				matching = append(matching, entry)
+			}
+		}
+	}
+	return matching
+}
+
+func (p *PitCsTree) FindInterestPrefixMatchByData(data *ndn.Data, token *uint32) []PitEntry {
+	if token != nil {
+		if entry, ok := p.pitTokenMap[*token]; ok && entry.Token() == *token {
+			return []PitEntry{entry}
+		}
+		return nil
+	}
+
+	matching := make([]PitEntry, 0)
```

Comment:
ACTION REQUIRED:
It's not good to copy code. You may write a common function (private visibility) to handle this.

Reply:
Done, the code shared by both `FindInterestPrefixMatch()` and `FindInterestPrefixMatchByData()` has been refactored out into `findInterestPrefixMatchByName()`.